Support Apple Pay in cross-origin iframes with allow=payment attribute

Zeswen opened the pull request:
#### fead01e13ad2f5837324865b19f30607383d410c
<pre>
Support Apple Pay in cross-origin iframes with allow=payment attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=226345">https://bugs.webkit.org/show_bug.cgi?id=226345</a>

Reviewed by Dean Jackson.

This bug happens because there is a check for the parentDocument to have
the same security origin than the child document, contrary to what the
[Payment Request API spec from W3 states](<a href="https://www.w3.org/TR/payment-request/#using-with-cross-origin-iframes).">https://www.w3.org/TR/payment-request/#using-with-cross-origin-iframes).</a>
The solution is to remove the check for the same parentDocument origin.

* LayoutTests/http/tests/paymentrequest/payment-allow-attribute.https-expected.txt:
- Update tests according to new behavior
* Source/WebCore/Modules/applepay/PaymentSession.cpp:
(PaymentSession::canCreateSession):
- Remove check for same parent origin since we have allow payment check before

Canonical link: <a href="https://commits.webkit.org/262616@main">https://commits.webkit.org/262616@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/29ebb3c627747050737e9853ebb83816771effb2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2048 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2076 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2143 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2973 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2111 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2022 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2156 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2118 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1871 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2066 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1838 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1839 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2823 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1824 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1815 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1728 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1884 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1863 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2998 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1881 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1682 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1808 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1826 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/511 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1988 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

marcoscaceres reviewed (commented):
thanks @Zeswen! 

This lgtm from a spec conformance standpoint. However, as I'm not a Webkit reviewer I can't give it a proper approval. I've pinged folks to find out if Apple Pay can also handle this and will get another person to review it soon!

grorg commented on Source/WebCore/Modules/applepay/PaymentSession.cpp:
Note to other folks looking: you need to expand the diff upwards to see that we test for `isFeaturePolicyAllowedByDocumentAndAllOwners` with `Featurepolicy::Type::Payment`.

grorg reviewed (approved):
This looks good. @geoffreygaren or someone should confirm.

Commit:
Support Apple Pay in cross-origin iframes with allow=payment attribute
https://bugs.webkit.org/show_bug.cgi?id=226345

Reviewed by Dean Jackson.

This bug happens because there is a check for the parentDocument to have
the same security origin than the child document, contrary to what the
[Payment Request API spec from W3 states](https://www.w3.org/TR/payment-request/#using-with-cross-origin-iframes).
The solution is to remove the check for the same parentDocument origin.

* LayoutTests/http/tests/paymentrequest/payment-allow-attribute.https-expected.txt:
- Update tests according to new behavior
* Source/WebCore/Modules/applepay/PaymentSession.cpp:
(PaymentSession::canCreateSession):
- Remove check for same parent origin since we have allow payment check before

Canonical link: https://commits.webkit.org/262616@main
